### PR TITLE
Global 5 tuple conntrack.

### PR DESCRIPTION
**Pull request by mchalla**
- support both global 5 tuple / container local 4 tuple conntrack
- cilium daemon can configure this support using #define CONNTRACK_LOCAL
- defaults to global

(Pending integration with Daniels changes and some more testing)
Signed-off-by: Madhu Challa <madhu@cilium.io>
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/cilium/cilium/pull/462%23issuecomment-289258406%22%2C%20%22https%3A//github.com/cilium/cilium/pull/462%23issuecomment-289308040%22%2C%20%22https%3A//github.com/cilium/cilium/pull/462%23issuecomment-289310001%22%2C%20%22https%3A//github.com/cilium/cilium/pull/462%23issuecomment-290226180%22%2C%20%22https%3A//github.com/cilium/cilium/pull/462%23discussion_r109049905%22%2C%20%22https%3A//github.com/cilium/cilium/pull/462%23discussion_r109050607%22%2C%20%22https%3A//github.com/cilium/cilium/pull/462%23issuecomment-290567458%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/cilium/cilium/pull/462%23issuecomment-289258406%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Still%20debugging%20a%20verifier%20issue.%20Not%20ready%20to%20merge.%22%2C%20%22created_at%22%3A%20%222017-03-26T04%3A51%3A18Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/725475%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mchalla%22%7D%7D%2C%20%7B%22body%22%3A%20%22Given%20that%20the%20key%20is%20now%20different%20between%20global%20and%20local%2C%20%60bpf/maps/ctmap%60%20must%20be%20adjusted%20as%20well.%22%2C%20%22created_at%22%3A%20%222017-03-26T19%3A23%3A19Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1417913%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/tgraf%22%7D%7D%2C%20%7B%22body%22%3A%20%22The%20verifier%20issue%20is%20resolved.%20I%20also%20forgot%20to%20change%20the%20code%20where%20the%20tuple%20is%20reversed.%20Added%20that.%20And%20will%20also%20adjust%20ctmap%20after%20integrating%20with%20Daniels%20change.%22%2C%20%22created_at%22%3A%20%222017-03-26T19%3A54%3A28Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/725475%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mchalla%22%7D%7D%2C%20%7B%22body%22%3A%20%22Added%20new%20keys%2C%20GC%20and%20Dump%20routines%20for%20the%20Global%20table.%20This%20needs%20Daniels%20changes.%20%5Cr%5Cnhttps%3A//github.com/cilium/cilium/commit/0b4dcbda36552bcb09474bcd96b94a75dc3e46e3%22%2C%20%22created_at%22%3A%20%222017-03-29T21%3A07%3A20Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/725475%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mchalla%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ok%2C%20looks%20like%20I%20need%20to%20downgrade%20my%20clang%20%603.9.1%60%20first%2C%20was%20running%20it%20for%20testing%20with%20newer%20ones.%20Since%20proxy%20changes%20on%20bpf%20side%20in%20current%20master%2C%20I%20get%20rejections%20with%20newer%20clang%20versions%20due%20to%20hitting%20complexity%20limit%20on%20some%20of%20the%20tail-called%20programs%20due%20to%20different%20code%20generation.%22%2C%20%22created_at%22%3A%20%222017-03-30T22%3A51%3A09Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/677393%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/borkmann%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20439a79a97da210cecfab91af765316065b7cf65c%20daemon/bindata.go%207%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/cilium/cilium/pull/462%23discussion_r109049905%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22For%20some%20reason%20your%20go-bindata%20contains%20%60a.sh%60%20and%20%60bpf_lxc.S%60.%20Looks%20like%20they%20are%20just%20local%20files%20from%20testing/debugging%3F%22%2C%20%22created_at%22%3A%20%222017-03-30T22%3A10%3A12Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/677393%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/borkmann%22%7D%7D%2C%20%7B%22body%22%3A%20%22yep%2C%20I%20should%20not%20include%20those.%20thanks.%22%2C%20%22created_at%22%3A%20%222017-03-30T22%3A14%3A34Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/725475%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mchalla%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20daemon/bindata.go%3AL1-10%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/cilium/cilium/pull/462#issuecomment-289258406'>General Comment</a></b>
- <a href='https://github.com/mchalla'></a> Still debugging a verifier issue. Not ready to merge.
- <a href='https://github.com/tgraf'></a> Given that the key is now different between global and local, `bpf/maps/ctmap` must be adjusted as well.
- <a href='https://github.com/mchalla'></a> The verifier issue is resolved. I also forgot to change the code where the tuple is reversed. Added that. And will also adjust ctmap after integrating with Daniels change.
- <a href='https://github.com/mchalla'></a> Added new keys, GC and Dump routines for the Global table. This needs Daniels changes.
https://github.com/cilium/cilium/commit/0b4dcbda36552bcb09474bcd96b94a75dc3e46e3
- <a href='https://github.com/borkmann'></a> Ok, looks like I need to downgrade my clang `3.9.1` first, was running it for testing with newer ones. Since proxy changes on bpf side in current master, I get rejections with newer clang versions due to hitting complexity limit on some of the tail-called programs due to different code generation.
- [ ] <a href='#crh-comment-Pull 439a79a97da210cecfab91af765316065b7cf65c daemon/bindata.go 7'></a> &nbsp;<b><a href='https://github.com/cilium/cilium/pull/462#discussion_r109049905'>File: daemon/bindata.go:L1-10</a></b>
- <a href='https://github.com/borkmann'></a> For some reason your go-bindata contains `a.sh` and `bpf_lxc.S`. Looks like they are just local files from testing/debugging?
- <a href='https://github.com/mchalla'></a> yep, I should not include those. thanks.

<a href='https://www.codereviewhub.com/cilium/cilium/pull/462?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/cilium/cilium/pull/462?approve=1'></a>&nbsp;<a href='https://github.com/cilium/cilium/pull/462'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by aanm** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/cilium/cilium/pull/462'>Refresh</a>.<a href='#crh-update-none'></a>

**Comment by mchalla** [comment]
Still debugging a verifier issue. Not ready to merge.

**Comment by tgraf** [comment]
Given that the key is now different between global and local, `bpf/maps/ctmap` must be adjusted as well.

**Comment by mchalla** [comment]
The verifier issue is resolved. I also forgot to change the code where the tuple is reversed. Added that. And will also adjust ctmap after integrating with Daniels change.

**Comment by mchalla** [comment]
Added new keys, GC and Dump routines for the Global table. This needs Daniels changes. 
https://github.com/cilium/cilium/commit/0b4dcbda36552bcb09474bcd96b94a75dc3e46e3

File: daemon/bindata.go [hunk 1]
```diff
@@ -1,7 +1,9 @@
 // Code generated by go-bindata.
 // sources:
 // ../bpf/COPYING
+// ../bpf/a.sh
 // ../bpf/bpf_lb.c
+// ../bpf/bpf_lxc.S
 // ../bpf/bpf_lxc.c
```
Comment (borkmann) [inline daemon/bindata.go:7]: For some reason your go-bindata contains `a.sh` and `bpf_lxc.S`. Looks like they are just local files from testing/debugging?

Comment (mchalla) [reply daemon/bindata.go:7]: yep, I should not include those. thanks.

**Comment by borkmann** [comment]
Ok, looks like I need to downgrade my clang `3.9.1` first, was running it for testing with newer ones. Since proxy changes on bpf side in current master, I get rejections with newer clang versions due to hitting complexity limit on some of the tail-called programs due to different code generation.

**Comment by tgraf** [comment]
@mchalla What's the current status of this?

**Comment by borkmann** [comment]
I have a rebase of this set and did various tests already. Will push it out today, so we can merge it and also close 443, too.

**Comment by borkmann** [comment]
Closing, handled further in pr #659 .

**Pull request closed without merging**